WalletWasabi/Crypto/Extensions.cs

```diff
@@ -1,11 +1,38 @@
 using System.Collections.Generic;
+using WalletWasabi.Helpers;
 using WalletWasabi.Crypto.Groups;
+using WalletWasabi.Crypto.ZeroKnowledge.LinearRelation;
+using WalletWasabi.Crypto;
 
 namespace System.Linq
 {
 	public static class Extensions
 	{
 		public static GroupElement Sum(this IEnumerable<GroupElement> groupElements) =>
 			groupElements.Aggregate(GroupElement.Infinity, (ge, acc) => ge + acc);
+
+		public static IEnumerable<TResult> Zip<TFirst, TSecond, TThird, TResult>(this IEnumerable<TFirst> first, IEnumerable<TSecond> second, IEnumerable<TThird> third, Func<TFirst, TSecond, TThird, TResult> resultSelector)
+		{
+			Guard.NotNull(nameof(first), first);
+			Guard.NotNull(nameof(second), second);
+			Guard.NotNull(nameof(third), third);
+			Guard.NotNull(nameof(resultSelector), resultSelector);
```

Comment:
Is this necessary? As https://github.com/zkSNACKs/WalletWasabi/issues/4241 mentions 

> Try to remove Guards from everywhere.

Is this code an exception from that rule?

Reply:
so my understanding is that since is only a warning so far Lucas prefers it at least in this type of code.

i looked into making these warnings into compile time errors, C# can do that but from what i understand only on a per project basis, not on a per file/assembly/namespace basis, but maybe it's possible?

Reply:
My point is that your IDE should show you when you attempt to pass nullable variable to a method parameter that accepts only non-null argument. (If you use MSVS or VS Code. Not sure about others.)

That analysis is actually so good that it's kind of hard to have NullReferenceException. So those guards, personally, do not help with anything. Glad to discuss this on Slack if we differ.

Also that argument of "this is crypto code then ..." I don't think it makes sense. Either those `Guards.NotNull` help to write better code or not. If they do then they should be used everywhere or they do not and they should not be used at all. Again, my personal view. But I find it weird to use `Guard.NotNull` here and not in the rest of the code. 

@nopara73 should probably say what he wants to align the approach everywhere and write proper conclusion here https://github.com/zkSNACKs/WalletWasabi/issues/4241.

Reply:
FWIW "this type" meant generic code, not crypto (that helper is just an extension to `IEnumerable<>` after all). although i committed this I didn't write it, hence deferring to @lontivero, but personally until a decision is made i would personally rather err on the side of paranoia, as it's pretty easy to remove and only hurts code consistency

related discussion: https://github.com/zkSNACKs/WalletWasabi/pull/4264#discussion_r481251113

as for IDE integration, i can confirm that in emacs w/ omnisharp these warnings are also displayed and are hard to miss. but when just using `dotnet` on the CLI, there are so many warnings that SNR is pretty bad right now, and since this is a bit unfamiliar to me this confused me and i ended up pushing bad code the other day because I missed errors ;-)

Reply:
I have to repeat myself here. We still don't know how this code will be used and we cannot prevent a caller from passing nulls so, if we decide to keep all this API private (currently it is public) then we can opt for removing the null checks. Nulls checks do not create defects and in the worst case they are just redundant or visual noise.